neon/src/connection.js

```diff
@@ -0,0 +1,1046 @@
+    chunk = nextWriteTimer = null
+    return x
+  }
+
+  function connectTimedOut() {
+    errored(Errors.connection('CONNECT_TIMEOUT', options, socket))
+    socket.destroy()
+  }
+
+  async function secure() {
+    write(SSLRequest)
+    const canSSL = await new Promise(r => socket.once('data', x => r(x[0] === 83))) // S
+
+    if (!canSSL && ssl === 'prefer')
+      return connected()
+
+    socket.removeAllListeners()
+    socket = tls.connect({
+      socket,
+      servername: net.isIP(socket.host) ? undefined : socket.host,
+      ...(ssl === 'require' || ssl === 'allow' || ssl === 'prefer'
+        ? { rejectUnauthorized: false }
+        : ssl === 'verify-full'
+          ? {}
+          : typeof ssl === 'object'
+            ? ssl
+            : {}
+      )
+    })
+    socket.on('secureConnect', connected)
+    socket.on('error', error)
+    socket.on('close', closed)
+    socket.on('drain', drain)
+  }
+
+  /* c8 ignore next 3 */
+  function drain() {
+    !query && onopen(connection)
+  }
+
+  function data(x) {
+    if (incomings) {
+      incomings.push(x)
+      remaining -= x.length
+      if (remaining >= 0)
+        return
+    }
+
+    incoming = incomings
+      ? Buffer.concat(incomings, length - remaining)
+      : incoming.length === 0
+        ? x
+        : Buffer.concat([incoming, x], incoming.length + x.length)
+
+    while (incoming.length > 4) {
+      length = incoming.readUInt32BE(1)
+      if (length >= incoming.length) {
+        remaining = length - incoming.length
+        incomings = [incoming]
+        break
+      }
+
+      try {
+        handle(incoming.subarray(0, length + 1))
+      } catch (e) {
+        query && (query.cursorFn || query.describeFirst) && write(Sync)
+        errored(e)
+      }
+      incoming = incoming.subarray(length + 1)
+      remaining = 0
+      incomings = null
+    }
+  }
+
+  async function connect() {
+    terminated = false
+    backendParameters = {}
+    socket || (socket = await createSocket())
+
+    if (!socket)
+      return
+
+    connectTimer.start()
+
+    if (options.socket)
+      return ssl ? secure() : connected()
+
+    socket.on('connect', ssl ? secure : connected)
+
+    if (options.path)
+      return socket.connect(options.path)
+
+    socket.ssl = ssl
+    socket.connect(port[hostIndex], host[hostIndex])
+    socket.host = host[hostIndex]
+    socket.port = port[hostIndex]
+
+    hostIndex = (hostIndex + 1) % port.length
+  }
+
+  function reconnect() {
+    setTimeout(connect, closedDate ? closedDate + delay - performance.now() : 0)
+  }
+
+  function connected() {
+    try {
+      statements = {}
+      needsTypes = options.fetch_types
+      statementId = Math.random().toString(36).slice(2)
+      statementCount = 1
+      lifeTimer.start()
+      socket.on('data', data)
+      keep_alive && socket.setKeepAlive && socket.setKeepAlive(true, 1000 * keep_alive)
+      const s = StartupMessage()
+      write(s)
+      AuthenticationCleartextPassword()
```

Comment:
This utilizes neon's support for "pipelining" these initial messages to the db. We send the cleartext password and ready for query events before the database actually requests those messages. This cuts down the number of round trips needed to the initialize the connection. 

You read more about this here: https://neon.tech/blog/quicker-serverless-postgres